### PR TITLE
add xor,xnor and tests

### PR DESCRIPTION
**Pull request by ranjithkumar007**

### CODE REVIEW

File: symengine/logic.cpp [hunk 1]
```diff
@@ -352,6 +352,20 @@ RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
...
 }
 
+RCP<const Boolean> logical_xor(const set_boolean &s)
+{
+    int cnttrue = 0;
+    for (auto &a : s) {
+        if (eq(*a, *boolTrue)) {
+            cnttrue++;
+        }
+    }
```
Comment (isuruf) [inline symengine/logic.cpp]: This function only checks for trues and false. `logical_xor(x, y)` would give false.

**Comment by ranjithkumar007** [comment]
there are two problems with the current implementation of logic module.
i) there is a lot of redundant code ...we can avoid this  by making a base class for and,or and xor similar to OneArgFunction for all trignometric functions.
ii)we can't pass symbols as arguments as it takes set_boolean as input param. 
@isuruf Suggestions?

**Comment by ranjithkumar007** [comment]
@isuruf  I can reproduce part 1 of my earlier comment if you agree !

File: symengine/logic.cpp [hunk 2]
```diff
@@ -352,6 +404,43 @@ RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
     }
 }
 
+RCP<const Boolean> logical_xor(const set_boolean &s)
```
Comment (isuruf) [inline symengine/logic.cpp]: If you use a `set` here, how would you pass `x, x`

Comment (ranjithkumar007) [inline symengine/logic.cpp]: need to change this....Same problem is with all other gates!

Comment (isuruf) [inline symengine/logic.cpp]: `Or` and `And` is fine.

Comment (ranjithkumar007) [inline symengine/logic.cpp]: symbols can be used as parameters if set_boolean is replaced by set_basic.

Comment (isuruf) [inline symengine/logic.cpp]: From `x` I meant the same boolean. for example, `y < 0, y < 0` where `y` is a symbol.

**Comment by ranjithkumar007** [comment]
@isuruf ,any idea why this [job](https://ci.appveyor.com/project/certik/symengine/build/2602/job/8p99aaeh7452ej4o) is failing?
Based on exec time [0.65s](https://ci.appveyor.com/project/certik/symengine/build/2602/job/8p99aaeh7452ej4o#L2470) , i feel that the cause might be an infinite recursion but unable to figure out exactly...!

File: symengine/logic.cpp [hunk 3]
```diff
@@ -352,6 +422,53 @@ RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
     }
 }
 
+RCP<const Boolean> logical_xor(const vec_boolean &s)
+{
+    set_boolean args;
+    for (auto &a : s) {
+        if (is_a<BooleanAtom>(*a)) {
+            auto val = down_cast<const BooleanAtom &>(*a).get_val();
+            if (val == false) {
```
Comment (isuruf) [inline symengine/logic.cpp]: Why is `false` ignored?

File: symengine/logic.cpp [hunk 4]
```diff
@@ -352,6 +422,53 @@ RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
...
+    if (args.size() == 0) {
+        return boolFalse;
+    } else if (args.size() == 1) {
+        return *args.begin();
+    } else {
+        for (auto &a : args) {
+            if (is_a<BooleanAtom>(*a)) {
+                SYMENGINE_ASSERT(a == boolTrue)
+                args.erase(a);
+                return logical_xnor(get_vec_from_set(args));
```
Comment (isuruf) [inline symengine/logic.cpp]: For performance reasons, these type of recursion should be avoided.

File: symengine/logic.cpp (hunk 3 above)
Comment (ranjithkumar007) [inline symengine/logic.cpp]: because Xor(args,false) = Xor(args)

File: symengine/logic.cpp (hunk 4 above)
Comment (ranjithkumar007) [inline symengine/logic.cpp]: yes.. i would convert it to iterative way..

File: symengine/logic.cpp [hunk 5]
```diff
@@ -448,24 +448,43 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+    while (args.size() > 1 and not cantsimplify) {
+        cantsimplify = true;
         for (auto &a : args) {
             if (is_a<BooleanAtom>(*a)) {
                 SYMENGINE_ASSERT(a == boolTrue)
                 args.erase(a);
-                return logical_xnor(get_vec_from_set(args));
+                cantsimplify = false;
+                nots++;
             } else if (args.find(logical_not(a)) != args.end()) {
```
Comment (isuruf) [inline symengine/logic.cpp]: This can be checked in Line 447 and don't need to do this here.

File: symengine/logic.cpp [hunk 6]
```diff
@@ -448,24 +448,43 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
-    } else {
+    while (args.size() > 1 and not cantsimplify) {
+        cantsimplify = true;
         for (auto &a : args) {
             if (is_a<BooleanAtom>(*a)) {
                 SYMENGINE_ASSERT(a == boolTrue)
                 args.erase(a);
-                return logical_xnor(get_vec_from_set(args));
+                cantsimplify = false;
+                nots++;
```
Comment (isuruf) [inline symengine/logic.cpp]: You can do this in the `if` condition in line 429

File: symengine/logic.cpp [hunk 7]
```diff
@@ -448,24 +448,43 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
     }
+    int nots = 0;
+    bool cantsimplify = false;
 
-    if (args.size() == 0) {
-        return boolFalse;
-    } else if (args.size() == 1) {
-        return *args.begin();
-    } else {
+    while (args.size() > 1 and not cantsimplify) {
```
Comment (isuruf) [inline symengine/logic.cpp]: why is this run multiple times?

File: symengine/logic.cpp [hunk 8]
```diff
@@ -448,24 +448,43 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+        cantsimplify = true;
         for (auto &a : args) {
             if (is_a<BooleanAtom>(*a)) {
                 SYMENGINE_ASSERT(a == boolTrue)
                 args.erase(a);
```
Comment (isuruf) [inline symengine/logic.cpp]: You are erasing from the vector while iterating it. Segfaults can happen if you do that.

**Comment by ranjithkumar007** [comment]
ping @isuruf

**Comment by ranjithkumar007** [comment]
ping @isuruf

**Comment by ranjithkumar007** [comment]
updated !

**Comment by ranjithkumar007** [comment]
@isuruf  , PR is ready for a final review !

**Comment by ranjithkumar007** [comment]
ping @isuruf

**Pull request merged**